### PR TITLE
Create concern to authenticate deep links using Turbolinks

### PR DESCRIPTION
**Pull request by NabeelAhsen**
> Related https://github.com/Shopify/shopify_app/issues/1092

This is related to supporting session tokens using Turbolinks. This PR introduces an optional controller concern that helps authenticate multi-page apps when accessing a link directly. 

Currently, with third party cookies disabled, a user is unable to access protected endpoints like `/products` or `/widgets`. When using Turbolinks, the recommended way of issuing a valid session token via App Bridge is by creating a splash page as the `root` resource of the app.

More info: https://shopify.dev/tutorials/authenticate-server-side-rendered-apps-with-session-tokens-app-bridge-turbolinks

This concern redirects an app to the `root` controller along with a `return_to` and `shop` parameter if it detects that an action is unauthenticated (i.e it fails to detect a valid session token session). Once a session token has been issued by App Bridge, Turbolinks can use this to redirect the app back to the `return_to` parameter in `shopify_app.js`:

```js
  // Wait for a session token before trying to load an authenticated page
  await retrieveToken(app);

  // Redirect to the requested page
  Turbolinks.visit(data.loadPath);
```

To use this concern, add it to the sample `AuthenticatedController` that gets generated by the Shopify App gem:

```ruby
# frozen_string_literal: true

class AuthenticatedController < ApplicationController
  include ShopifyApp::TurbolinksAuthenticated # <= Add this here
  include ShopifyApp::Authenticated
end
```

- [x] Show this works using app extensions and direct app-links
- [ ] Update `CHANGELOG.md` if the changes would impact users
- [x] Update `README.md`, if appropriate.
- [x] Update any relevant pages in `docs/`, if necessary
- [x] For security fixes, the [Disclosure Policy](https://github.com/Shopify/shopify_app/blob/master/SECURITY.md#disclosure-policy) must be followed.

### CODE REVIEW

File: app/controllers/concerns/shopify_app/turbolinks_authenticated.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module ShopifyApp
+  module TurbolinksAuthenticated
```
Comment (NabeelAhsen) [inline app/controllers/concerns/shopify_app/turbolinks_authenticated.rb:4]: I'm not entirely sure what to call this concern yet, I'm open to suggestions!

**Comment by thecodepixi** [comment]
Hey @NabeelAhsen, this looks awesome, but it looks like there are 2 small linter errors at the moment.

**Comment by NabeelAhsen** [comment]
> Hey @NabeelAhsen, this looks awesome, but it looks like there are 2 small linter errors at the moment.

Oh yeah I've been trying to run `bundle exec rubocop` locally to fix these, but to no avail! It seems that there's a mismatch in the rubocop versions in the CI vs my local environment 😬 

Do you have any recommendations around this?

**Comment by mllemango** [comment]
Sorry that's on me 😅, PR's' up!

**Comment by thecodepixi** [comment]
I just noticed the same. It looks like that rubocop error exists on `master`, so you might want to rebase on `master` when that fix is merged. 

As for why `rubocop` isn't working for you locally, it looks like the CI just runs the latest version of the gems (`rubocop` and `rubocop-shopify`) so you may just need to update them locally as well.

**Comment by thecodepixi** [comment]
> Sorry that's on me 😅, PR's' up!

Approved ✅

**Comment by thecodepixi** [comment]
@NabeelAhsen You should be good to rebase `master` to get rid of that `rubocop` error 👍

**Comment by andyw8** [comment]
I opened https://github.com/Shopify/shopify_app/pull/1120 which should help to avoid some types of RuboCop version mismatch.

**Review by ragalie — APPROVED** [review]
Makes sense to me!

**Review by mSitkovets — APPROVED** [review]

**Review by paulomarg — APPROVED** [review]

**Comment by NabeelAhsen** [comment]
I've added a section in `README.md` to explain this concern a little better. I'm still having a hard time naming this thing -- does `EnsureAuthenticatedLinks` sound too confusing?

**Review by paulomarg — APPROVED** [review]
I'm not very good at naming things, but I kind of like what we have. The README text LGTM.

File: README.md [hunk 2]
```diff
@@ -279,6 +279,24 @@ The engine provides a `ShopifyApp::Authenticated` concern which should be includ
 
 For backwards compatibility, the engine still provides a controller called `ShopifyApp::AuthenticatedController` which includes the `ShopifyApp::Authenticated` concern. Note that it inherits directly from `ActionController::Base`, so you will not be able to share functionality between it and your application's `ApplicationController`.
 
+### EnsureAuthenticatedLinks
+
+The `ShopifyApp::EnsureAuthenticatedLinks` concern helps authenticate users that access protected pages of your app directly.
+
+Include this concern in your app's `AuthenticatedController` if your app uses session tokens with [Turbolinks](https://shopify.dev/tutorials/authenticate-server-side-rendered-apps-with-session-tokens-app-bridge-turbolinks). It adds a `before_action` filter that detects whether a session token is present or not. If a session is not found, the app is redirected to your app's splash page path (`root_path`) along with `return_to` and `shop` parameters.
```
Comment (andyw8) [inline README.md:286]: "the app is redirected" - should that say "the user is redirected" ?

File: README.md [hunk 3]
```diff
@@ -279,6 +279,24 @@ The engine provides a `ShopifyApp::Authenticated` concern which should be includ
...
+
+Example `AuthenticatedController`:
+
+```rb
+# frozen_string_literal: true
```
Comment (andyw8) [inline README.md:291]: I don't think we need to include the `frozen_string_literal` line.

File: README.md (hunk 2 above)
Comment (NabeelAhsen) [reply README.md:286]: Good catch!

**Review by rezaansyed — APPROVED** [review]
LGTM 👍

**Review by greatestape — APPROVED** [review]
This is great! Nicely done.

**Review by mSitkovets — APPROVED** [review]

**Comment by netwire88** [comment]
Hi @NabeelAhsen, I noticed that the order of the concerns matter.

This doesn't work. If you go to a deep link in an incognito browser window (or any browser window not logged into Shopify), you get an `ShopifyDomainNotFound` error.

```
class AuthenticatedController < ApplicationController
  include ShopifyApp::EnsureAuthenticatedLinks
  include ShopifyApp::Authenticated
end
```

However, if I change it to this, it works. It'll ask for myshopify URL first and then login in + redirect.

```
class AuthenticatedController < ApplicationController
  include ShopifyApp::Authenticated
  include ShopifyApp::EnsureAuthenticatedLinks
end
```

Can you confirm?

**Comment by NabeelAhsen** [comment]
> Hi @NabeelAhsen, I noticed that the order of the concerns matter.
> 
> This doesn't work. If you go to a deep link in an incognito browser window (or any browser window not logged into Shopify), you get an `ShopifyDomainNotFound` error.
> 
> ```
> class AuthenticatedController < ApplicationController
>   include ShopifyApp::EnsureAuthenticatedLinks
>   include ShopifyApp::Authenticated
> end
> ```
> 
> However, if I change it to this, it works. It'll ask for myshopify URL first and then login in + redirect.
> 
> ```
> class AuthenticatedController < ApplicationController
>   include ShopifyApp::Authenticated
>   include ShopifyApp::EnsureAuthenticatedLinks
> end
> ```
> 
> Can you confirm?

@netwire88 Thank you for bringing this to our attention! This is definitely an oversight, and is being fixed with this PR: https://github.com/Shopify/shopify_app/pull/1158

The PR i linked here will allow the order of `EnsureAuthenticatedLinks` and `Authenticated` to stay the same.

**Comment by netwire88** [comment]
Thanks @NabeelAhsen !

**Pull request merged**